Remove undefined behavior

lpha-z opened the pull request:
ブランチを分け、目についた未定義動作を取り除きました。
また、コメントを充実させました。

lpha-z committed:
Fix CorrectEndian() usage

lpha-z committed:
Fix ExtractBits() overflow bug

The type of ~0<<len is int; this is incorrect when T = u64.
A simple fix ~(T)0<<len overflows when len=64.
Although it is unclear whether it is a function assuming input of
len=64, I chose the safety side.

shioyadan commented:
修正ありがとう．
ちょっといくつかまだコメントがあるので，以下に書きます．

SafeAbs の現在のコメントは「今の実装がどのように働き，なぜ問題が生じないのか」という視点で書かれていて，実装そのものは結構トリッキーなものになってると思う（つまり，この実装がどのように働くかを丁寧に説明すると，このぐらいの長さのコメントがいるってコト自体がそれを表していると思う）．

なので，コード自体を，誰が読んでもわかるようにわかる形に書いてほしくて，たとえばそれは
    
    if (x == s64_MIN) {
        // この値の場合，単純に符号反転すると s64 では表せなくてオーバーフローするから特別処理する YO とかそういうコメント
        return minus_s64_MIN; （なんか s64_MIN を符号反転した定数）
    }
    else {
        return (u64)(x < 0 ? -x : x);
    }
のように書けば，ビット演算の細かいルールを理解しなくても良くなって，良いと思う．
（要は，コメントをそもそも書かずにすむようにコードが書ければその方が良いというはなし．この辺の話は，Code Complete って本がそのへんに落ちてると思うので，自己言及可能なコードとかそういうあたりの章を，読んだことなかったら読んでみると参考になるかもしんない）

AsIntFunc の部分については，修正後のコードだけ見ると union の影も形もないので，修正後のコードだけ見た人には何いってんだかよくわからんコメントになると思う．あと，「strict aliase rules に反するので」とだけ書くよりは，union を使って ビット表現を再解釈することは C++ の規格では基本的に許されないってところまで説明したほうが親切だと思う．

あとこれ，sizeof にカッコがついてないけど，コンパイル通ってる？

lpha-z commented:
* まず、`s64`は常に二の補数で書き表されていると仮定してよいのでしょうか。その制約が明示されていないせいで冗長なコメントが必要になっていますし、`minus_s64_MIN`のような値の決め打ちも書けなくなっています。
* `union`の説明は確かに舌足らずな感じがあるので、増やします。
* `sizeof`演算子は前置単項演算子なので、`sizeof x`のような書き方は`-x`同様に正しい書き方です。かっこが必須なのは、`sizeof(Type)`の場合です。多くのプログラマは`sizeof(x)`と書くようですが。

shioyadan commented:
修正いろいろおつかれさまです，ありがとう．

>・まず、s64は常に二の補数で書き表されていると仮定してよいのでしょうか。その制約が明示されていないせいで冗長なコメントが必要になっていますし、minus_s64_MINのような値の決め打ちも書けなくなっています。

これはさすがに二の補数は仮定して良いと思う．エミュレーション対象の各 ISA も符号付き演算は二の補数が前提になっているわけで，ホスト側が二の補数じゃないマシンだったら，色んな部分でたぶんそもそも動かんかなと．あと，二の補数じゃないマシン，現在は俺らが使うことあるかといえば事実上使うことはないだろうし，今後も出てこないと思う．

> sizeof演算子は前置単項演算子なので、sizeof xのような書き方は-x同様に正しい書き方です。かっこが必須なのは、sizeof(Type)の場合です。多くのプログラマはsizeof(x)と書くようですが。

あー，そうなんだ，それは知らなかった．そこはちょっと，他の部分とのコードの一貫性を保つために，カッコ付きで書いておいてもらえると助かるかも．


>    // UnsignedMulHigh64の引数は高々二の三十一乗である。かけ合わせた結果は
>    //  高々二の六十二乗であり、正負ともに余裕をもってs64で表せる

ここのコメントがちょっとわからなかったんだけど，UnsignedMulHigh64 の内部では引数を 2 の 31 乗だとは思ってなくない？

lpha-z commented:
ちょっとした勘違いで間違ったコメントをつけていたみたいです。直しておきます

lpha-z committed:
Remove undefined behavior (strict aliasing rules) in AsFPFunc() and AsIntFunc()

lpha-z committed:
Remove undefined behavior (signed integer overflow) in SignedMulHigh64()

Improve the prospect of dealing with overflow by using the integer type
represented by *always* two's complement

lpha-z commented:
指摘は全部反映させたと思います。これでどうでしょうか。

shioyadan commented:
確認してみたけど，大丈夫だと思う．ありがとう．
ちょっと実際に走らしてみて最終確認してから master にマージするので，マージ自体はもうしばらくお待ちを．

shioyadan commented:
SPEC CPU2006 で検証したところ，びみょーに変わったりすることもあるだろかと予想してたんだけど，実際はもろもろのマージ前後で一切 IPC が変化しとりませんでした．

問題はないということで，全部マージしました．
ありがとう．
